machine-check/src/swarm.rs

```diff
@@ -0,0 +1,729 @@
+    let empty = BTreeSet::new(); // just for `sub` but needs its own lifetime
+    let sub = |r: &Role| subs.get(r).unwrap_or(&empty);
+
+    // visit all reachable nodes of the graph to check their prescribed conditions; order doesn’t matter
+    for node in Dfs::new(&graph, initial).iter(&graph) {
+        let mut guards = BTreeMap::new();
+        let mut commands = BTreeSet::new();
+        for edge in graph.edges_directed(node, Outgoing) {
+            let log = edge.weight().log_type.as_slice();
+
+            // event determinism
+            let guard = &log[0];
+            if *guards
+                .entry(guard.clone())
+                .and_modify(|count| *count += 1)
+                .or_insert(1)
+                == 2
+            {
+                errors.push(Error::NonDeterministicGuard(edge.id()));
+            }
+            // command determinism
+            let command = &edge.weight().cmd;
+            let role = &edge.weight().role;
+            if !commands.insert((role.clone(), command.clone())) {
+                errors.push(Error::NonDeterministicCommand(edge.id()));
+            }
+
+            let target = edge.target();
+
+            // causal consistency
+            if log_filter(log, sub(&role)).first_one().is_none() {
+                errors.push(Error::ActiveRoleNotSubscribed(edge.id()));
+            }
+            for active in &graph[target].active {
+                let filtered = log_filter(log, sub(active));
+                if filtered.first_one().is_none() {
+                    errors.push(Error::LaterActiveRoleNotSubscribed(
+                        edge.id(),
+                        active.clone(),
+                    ));
+                }
+                for later in &graph[target].roles {
+                    let later_log = log_filter(log, sub(later));
+                    let extra = later_log & !filtered.clone();
+                    if extra.first_one().is_some() {
+                        errors.push(Error::LaterInvolvedRoleMoreSubscribed {
+                            edge: edge.id(),
+                            later: later.clone(),
+                            active: active.clone(),
+                            events: extra.iter_ones().map(|i| log[i].clone()).collect(),
+                        });
+                    }
+                }
+            }
+
+            // choice determinacy
+            for later in &graph[target].roles {
+                if !sub(later).contains(guard) {
+                    errors.push(Error::LaterInvolvedNotGuarded(edge.id(), later.clone()));
+                }
+            }
+        }
+    }
+    errors
+}
+
+pub fn from_json(
+    proto: SwarmProtocol,
+    subs: &Subscriptions,
+) -> (super::Graph, Option<NodeId>, Vec<String>) {
+    let (g, i, e) = prepare_graph(proto, subs);
+    (to_swarm(&g), i, e.map(Error::convert(&g)))
+}
+
+/// unfortunately there is no walker for neighbors, so we need to handroll it
```

Comment:
Does this one not work? https://docs.rs/petgraph/latest/petgraph/graph/struct.Neighbors.html